design/list_command_design.md

```diff
@@ -0,0 +1,266 @@
+         {list,variable,policy} ...
+conjur-cli: error: unrecognized arguments: boo
+
+## Should be (ignoring the last unknown input):
+[
+  "myorg:user:foo",
+  "myorg:variable:foo/bar"
+]
+
+```
+
+*Mispell list command*
+
+```bash
+$ conjur-cli listboo
+
+## Current
+usage: conjur-cli [-h] [-v] [-l URL] [-a ACCOUNT] [-c CA_BUNDLE] [--insecure] [-u LOGIN_ID] [-k API_KEY] [-p PASSWORD] [-d]
+                  [--verbose]
+                  {list,variable,policy} ...
+conjur-cli: error: argument resource: invalid choice: 'listl' (choose from 'list', 'variable', 'policy')
+
+## Should be:
+Error: Unknown command 'listboo'
+
+<Main help screen>
+```
+
+*Help flag*
+
+```bash
+$ conjur-cli list --help
+
+## Current
+usage: conjur-cli list [-h]
+
+optional arguments:
+  -h, --help  show this help message and exit
+  
+## Should be
+<List help screen>
+```
+
+*`--inspect` should show full object information*
+
+```bash
+$ conjur-cli list --inspect
+[
+		"policy_versions": [
+   	 {
+      "version": 1,
+      "created_at": "2020-11-17T12:46:51.070+00:00",
+      "policy_text": "- !policy\n  id: foo\n  body:\n    - &variables\n      - !variable bar \n",
+      "policy_sha256": "4b6fd7cfc622dd8790fd437390f5b079c0cd77e72aa862384c5b740bdcedb3c6",
+      "finished_at": "2020-11-17T12:46:51.311+00:00",
+      "client_ip": "172.17.0.1",
+      "id": "myorg:policy:root",
+      "role": "myorg:user:admin"
+    	},
+   	]
+  {
+    "created_at": "2020-11-18T12:00:00.078+00:00",
+    "id": "myorg:variable:secrets/test_secret",
+    "owner": "myorg:policy:secrets",
+    "policy": "myorg:policy:root",
+    "permissions": [
+
+    ],
+    "annotations": {
+    },
+    "secrets": [
+      {
+        "version": 1,
+        "expires_at": null
+      }
+    ]
+  },
+  {
+    "created_at": "2020-11-18T14:46:17.172+00:00",
+    "id": "myorg:user:boo",
+    "owner": "myorg:user:admin",
+    "policy": "myorg:policy:root",
+    "permissions": [
+
+    ],
+    "annotations": {
+    },
+    "restricted_to": [
+
+    ]
+  }
+]
+```
+
+*`--kind=user` gives use all user kinds*
+
+```bash
+conjur list --kind=user
+[
+  "myorg:user:boo"
+]
+```
+
+*`-k user` gives use all user kinds*
+
+```bash
+conjur list -k user
+[
+  "myorg:user:boo"
+]
+```
+
+*`--kind=user` gives us no user if no hosts exist*
+
+```bash
+conjur list --kind=user
+[
+
+]
+```
+
+*`--kind=boo` gives us no nothing if no that kind doesn't exist*
+
+```bash
+conjur list --kind=boo
+[
+
+]
+```
+
+*`--limit=4` returns top 4 from the list*
+
+```bash
+$ conjur list --limit=4
+[
+  "myorg:policy:foo",
+  "myorg:policy:root",
+  "myorg:policy:secrets",
+  "myorg:policy:users"
+]
+```
+
+*`-l 4` returns top 4 from the list*
+
+```bash
+$ conjur list -l 4
+[
+  "myorg:policy:foo",
+  "myorg:policy:root",
+  "myorg:policy:secrets",
+  "myorg:policy:users"
+]
+```
+
+*`--limit=-1` (negative number or 0) gives us the full list.* In the Ruby CLI, this is an 500 Internal Server Error
+
+```bash
+$ conjur list --limit=-1
```

Comment:
I think an error should be the result.
As a user i want to know that i add wrong input (for example in automation if i will do by mistake -1), the user may think its working when he have 1 item (it will show all the list) but then when we will have another item he will not understand why he get two items instead of 1.

Reply:
Agree with @eladkug, as showing the full list will draw too much of the user's attention from the actual error.

Reply:
@sigalsax what happens in the REST API?

Reply:
We can go with @eladkug suggestion and try it out in our UX tests

Reply:
@InbalZilberman the REST API returns a `HTTP/1.1 500 Internal Server Error`